File: customer/src/test/java/br/com/accenture/wallet/CustomerControllerTest.java
package br.com.accenture.wallet;

import br.com.accenture.wallet.customer.controller.CustomerController;
import br.com.accenture.wallet.customer.domain.CustomerModel;
import br.com.accenture.wallet.customer.service.CustomerService;
import org.junit.Before;
import org.junit.Test;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;
import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.Matchers.*;

public class CustomerControllerTest {

    private CustomerService mockService = mock(CustomerService.class);

    private CustomerController controller;

    @Before
    public void setUp() {
        controller = new CustomerController(mockService);
    }

    @Test
    public void testCreateCustomer() {
        CustomerModel model = new CustomerModel("", "")
            .setName("Abdoral Test")
            .setEmail("adboral.test@test.com")
            .setGender("male");

        when(mockService.create(model)).thenReturn(model.setId("abc123"));
        final CustomerModel modelResult = controller.create(model);
        assertThat(modelResult.getId(), notNullValue());
    }

    @Test
    public void testUpdateCustomer() {
        CustomerModel model = new CustomerModel("", "")
            .setName("Abdoral Test")
            .setEmail("adboral.test@test.com")
            .setGender("male");

        when(mockService.update(model)).thenReturn(model.setId("bcad-9173"));
        final CustomerModel modelResult = controller.update("bcad-9173", model);
        assertThat(modelResult.getId(), notNullValue());
    }

}

File: customer/src/test/kotlin/br/com/accenture/wallet/customer/CustomerControllerTest.kt
package br.com.accenture.wallet.customer

import br.com.accenture.wallet.customer.controller.CustomerController
import br.com.accenture.wallet.customer.domain.CustomerBean
import br.com.accenture.wallet.customer.service.CustomerService
import org.mockito.Mockito.mock
import org.mockito.Mockito.`when`
import org.junit.Before
import org.junit.Test
import org.hamcrest.MatcherAssert.assertThat
import org.hamcrest.Matchers.*
import javax.validation.Validation

class CustomerControllerTest {
    private val mockService = mock(CustomerService::class.java)

    lateinit var controller: CustomerController

    @Before
    fun setUp() {
        controller = CustomerController(mockService)
    }

    @Test
    fun testCreateCustomer() {
        val model = CustomerBean()
        model.name = "Abdoral Test"
        model.email = "adboral.test@test.com"
        model.gender = "male"

        `when`(mockService.create(model)).thenReturn(model)
        val result = controller.create(model)
        assertThat(result.id, notNullValue())
    }

    @Test
    fun testCreateCustomerWithValidation() {
        val validFactory = Validation.buildDefaultValidatorFactory()

        val model = CustomerBean()
        model.name = "Abdoral Test"

        val constraints = validFactory.validator.validate(model)

        assertThat(constraints.size, greaterThan(0))
    }

    @Test
    fun testUpdateCustomer() {
        val model = CustomerBean()
        model.name = "Abdoral Test"
        model.email = "adboral.test@test.com"
        model.gender = "male"

        `when`(mockService.update(model)).thenReturn(model)
        val result = controller.update("bcad-9173", model)
        assertThat(result?.id, notNullValue())
    }

}